crates/polars-ops/src/series/ops/cum_agg.rs

```diff
@@ -216,15 +216,37 @@ pub fn cum_max(s: &Series, reverse: bool) -> PolarsResult<Series> {
 }
 
 pub fn cum_count(s: &Series, reverse: bool) -> PolarsResult<Series> {
-    if reverse {
-        let ca: NoNull<UInt32Chunked> = (0u32..s.len() as u32).rev().collect();
-        let mut ca = ca.into_inner();
-        ca.rename(s.name());
-        Ok(ca.into_series())
-    } else {
-        let ca: NoNull<UInt32Chunked> = (0u32..s.len() as u32).collect();
-        let mut ca = ca.into_inner();
-        ca.rename(s.name());
-        Ok(ca.into_series())
+    // Fast paths for no nulls
+    if s.null_count() == 0 {
+        let out = cum_count_no_nulls(s.name(), s.len(), reverse);
+        return Ok(out);
     }
+
+    let mut count = 0 as IdxSize;
+    let f = |v: bool| {
+        if v {
+            count += 1
+        }
+        count
+    };
+
+    let ca = s.is_not_null();
```

Comment:
After this I think we should use `try_apply_values_generic`. That take a `FnMut` so should be able to increment the counter. It will ensure we iterate the data faster.

Reply:
That copies over the validity buffer, which we don't want. Is there a version that doesn't? 🤔

Reply:
I went with `apply_values_generic` as the closure doesn't return a Result.

The double reverse when `reverse` is true seems inefficient though.

Reply:
@stinodego as I mentioned above I believe you can start the counter at `s.len() - s.null_count()` and decrement the counter in the reverse case to avoid the double reverse...unless I'm misinterpreting how the `reverse` parameter works.

Reply:
That would lead to incorrect results.

Reply:
Hi @stinodego,

I spent a bit of time since I *knew* there had to be an algorithm that worked without the second rev. Here it is--our first element is guaranteed to be `s.len() - s.null_count()`, and the only difference from the normal accumulation is that if we see a null value, we don't decrement the `next` value, not the current one. Here's an example with the algorithm and a fair number of tests (in python):

```python
import polars as pl


def cum_count_rev(s):
    """Standard polars impl."""
    return pl.Series(s).fill_null(0).reverse().cum_sum().reverse().to_list()


def cum_count_rev2(s):
    """Impl with no second reversal."""
    s = pl.Series(s)  # single reversal required

    count = s.len() - s.null_count()
    out = []
    decr = 0
    for i in range(s.len()):
        count = count - decr
        out.append(count)

        # decide whether or not to decrement the next value
        decr = 0 if s[i] is None else 1

    return out


# test a couple examples
s_values = [
    [1, None, 1, 1, 1, None, None, None, 1, 1, 1, None, None],
    [None, 1, None, 1, None, None, None, 1, 1, 1, 1, 1, 1, 1],
    [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
    [None, None, None, None, None, None, None, None, None],
    [None, 1, 1, 1, 1, 1, 1, None, None, None, None],
]

for s in s_values:
    print(f"                s : {s}")
    print(f"cum_count_rev(s)  : {cum_count_rev(s)}")
    print(f"cum_count_rev2(s) : {cum_count_rev2(s)}")
```
```
                s : [1, None, 1, 1, 1, None, None, None, 1, 1, 1, None, None]
cum_count_rev(s)  : [7, 6, 6, 5, 4, 3, 3, 3, 3, 2, 1, 0, 0]
cum_count_rev2(s) : [7, 6, 6, 5, 4, 3, 3, 3, 3, 2, 1, 0, 0]
                s : [None, 1, None, 1, None, None, None, 1, 1, 1, 1, 1, 1, 1]
cum_count_rev(s)  : [9, 9, 8, 8, 7, 7, 7, 7, 6, 5, 4, 3, 2, 1]
cum_count_rev2(s) : [9, 9, 8, 8, 7, 7, 7, 7, 6, 5, 4, 3, 2, 1]
                s : [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
cum_count_rev(s)  : [14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
cum_count_rev2(s) : [14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
                s : [None, None, None, None, None, None, None, None, None]
cum_count_rev(s)  : [0, 0, 0, 0, 0, 0, 0, 0, 0]
cum_count_rev2(s) : [0, 0, 0, 0, 0, 0, 0, 0, 0]
                s : [None, 1, 1, 1, 1, 1, 1, None, None, None, None]
cum_count_rev(s)  : [6, 6, 5, 4, 3, 2, 1, 0, 0, 0, 0]
cum_count_rev2(s) : [6, 6, 5, 4, 3, 2, 1, 0, 0, 0, 0]
```

Reply:
Good one @mcrumiller!